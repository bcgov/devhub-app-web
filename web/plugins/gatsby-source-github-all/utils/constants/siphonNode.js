/*
Copyright 2018 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/

const { SOURCE_TYPES } = require('./sourceTypes');

const UNFURL_TYPES = {
  MARKDOWN: 'markdown',
  EXTERNAL: 'external',
};
// source types that map to fetch routines
const TOPIC_TYPES = {
  [SOURCE_TYPES.GITHUB]: 'source',
  [SOURCE_TYPES.WEB]: 'source',
  CURATED: 'curated',
};

const RESOURCE_TYPES = {
  COMPONENTS: 'Components',
  DOCUMENTATION: 'Documentation',
  SELF_SERVICE_TOOLS: 'Self-Service Tools',
  REPOSITORIES: 'Repositories',
  PEOPLE: 'People',
  PROJECTS: 'Projects',
  EVENTS: 'Events',
};
// Resource Types for Devhub
const RESOURCE_TYPES_LIST = Object.values(RESOURCE_TYPES);

const PERSONAS = {
  DESIGNER: 'Designer',
  DEVELOPER: 'Developer',
  PRODUCT_OWNER: 'Product Owner',
};

const TOPIC_TEMPLATES = {
  DEFAULT: 'default',
  OVERVIEW: 'overview',
};

const TOPIC_TEMPLATES_LIST = Object.values(TOPIC_TEMPLATES);

const PERSONAS_LIST = Object.values(PERSONAS);

const GRAPHQL_NODE_TYPE = {
  SIPHON: 'DevhubSiphon',
  TOPIC: 'DevhubTopic',
};

module.exports = {
  UNFURL_TYPES,
  TOPIC_TYPES,
  RESOURCE_TYPES,
  RESOURCE_TYPES_LIST,
  PERSONAS,
  PERSONAS_LIST,
  GRAPHQL_NODE_TYPE,
  TOPIC_TEMPLATES,
  TOPIC_TEMPLATES_LIST,
};
