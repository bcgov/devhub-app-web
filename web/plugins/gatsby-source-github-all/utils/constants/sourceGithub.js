/*
Copyright 2018 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/
const DEFUALT_IGNORES = [
  '/node_modules',
  'CONTRIBUTING.md',
  'CODE_OF_CONDUCT.md',
  'LICENSE',
  'CODE OF CONDUCT.md',
  '/openshift',
  '/.github',
  '.github',
];
// github rest api v3
const GITHUB_API_ENDPOINT = 'https://api.github.com';

const FILETYPES = {
  yml: 'YAML',
  yaml: 'YAML',
  json: 'JSON',
  md: 'Markdown',
  txt: 'text',
  html: 'HTML',
  HTML: 'HTML',
};

const MEDIATYPES = {
  md: 'text/markdown',
  json: 'application/json',
  yaml: 'text/yaml',
  yml: 'text/yaml',
  txt: 'text/html',
  html: 'text/html',
  HTML: 'text/html',
};

const PROCESSABLE_EXTENSIONS = ['.md'];

const MARKDOWN_FRONTMATTER_SCHEMA = {
  title: {
    type: String,
    required: false,
  },
  description: {
    type: String,
    required: true,
  },
  resourcePath: {
    type: String,
    required: false,
  },
  slug: {
    type: String,
    required: false,
  },
  ignore: {
    type: Boolean,
    required: false,
  },
  image: {
    type: String,
    required: false,
  },
  label1: {
    type: String,
    required: false,
  },
  data1: {
    type: String,
    required: false,
  },
  label2: {
    type: String,
    required: false,
  },
  data2: {
    type: String,
    required: false,
  },
  author: {
    type: String,
    required: false,
  },
  pageOnly: {
    type: Boolean,
    required: false,
  },
  resourceType: {
    type: String,
    required: false,
  },
  persona: {
    type: String,
    required: false,
  },
};

module.exports = {
  DEFUALT_IGNORES,
  GITHUB_API_ENDPOINT,
  FILETYPES,
  MEDIATYPES,
  PROCESSABLE_EXTENSIONS,
  MARKDOWN_FRONTMATTER_SCHEMA,
};
