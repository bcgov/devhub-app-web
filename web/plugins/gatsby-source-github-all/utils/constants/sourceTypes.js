/*
Copyright 2018 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/

const { isString, isArray, every } = require('lodash');
const messages = require('../console');
const SOURCE_TYPES = {
  GITHUB: 'github',
  WEB: 'web',
};

/**
 * checks to see if source properties contains correct config for source type github
 * @param {Object} properties the source properties from the registry
 */
const hasFileOrFiles = properties => {
  if (properties.file) {
    return isString(properties.file);
  } else if (!!properties.files) {
    return isArray(properties.files) && every(properties.files, isString);
  }
  // eslint-disable-next-line no-console
  console.warn(messages.deprecatedGithubSourceValidation());
  return false;
};

const GITHUB_SOURCE_SCHEMA = {
  url: {
    type: String,
    required: true,
  },
  owner: {
    type: String,
    required: true,
  },
  repo: {
    type: String,
    required: true,
  },
  file: {
    type: String,
    required: true,
    validate: hasFileOrFiles,
  },
  files: {
    type: Array,
    required: true,
    validate: hasFileOrFiles,
  },
  // branch to base fetch off of
  branch: {
    type: String,
    required: false,
  },
  // inline devhubignores
  ignores: {
    type: Array,
    required: false,
  },
};

const WEB_SOURCE_SCHEMA = {
  url: {
    type: String,
    required: true,
  },
  author: {
    type: String,
    required: false,
  },
  image: {
    type: String,
    required: false,
  },
  description: {
    type: String,
    required: false,
  },
  title: {
    type: String,
    required: false,
  },
  data1: {
    type: String,
    required: false,
  },
  label1: {
    type: String,
    required: false,
  },
  data2: {
    type: String,
    required: false,
  },
  label2: {
    type: String,
    required: false,
  },
};

module.exports = {
  SOURCE_TYPES,
  GITHUB_SOURCE_SCHEMA,
  WEB_SOURCE_SCHEMA,
};
