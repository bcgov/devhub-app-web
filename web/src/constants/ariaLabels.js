export const ARIA_LABEL_REPO = 'Link to the Github Repository';
export const ARIA_LABEL_FILE_IN_REPO =
  "Link to original source for this content found in it's Github Repository";
export const ARIA_LABEL_RESOURCE = 'Learn more about this article';
export const ARIA_LABEL_FILTER_SELECT = 'Select to filter cards by the resource type';
export const ARIA_LABEL_TO_GITHUB_USER = 'Select to view this authors Github Profile';
export const ARIA_LABEL_TO_GITHUB_ISSUES = "Select to view this Github respositories' issues";
export const ARIA_LABEL_WATCH_GITHUB_REPO = 'Watch this Github repository';
export const ARIA_LABEL_FORK_GITHUB_REPO = 'Fork this Github repository';
export const ARIA_LABEL_FILTER_RESOURCE = 'Filter resources by this value';
export const ARIA_LABEL_RESOURCES_MATCH_FILTER = 'Number of resources that match this filter';
// SEARCHING
export const ARIA_LABEL_SEARCH_BUTTON = 'Click to Search';
export const ARIA_LABEL_SEARCH_INPUT = 'Search input field';
