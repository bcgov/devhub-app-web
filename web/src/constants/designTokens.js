import { RESOURCE_TYPES, TOPICS, SEARCH_RESOURCE_TYPES } from './ui';

export const STYLE_BASE = {
  fontSize: 14,
  spacing: 5,
  borderWidth: 1,
};

// custom font size typing
// CUSTOM_TYPE.sm => 16px;
export const CUSTOM_TYPE = {
  sm: `${STYLE_BASE.fontSize}px`,
  md: `${STYLE_BASE.fontSize * 1.5}px`,
  lg: `${STYLE_BASE.fontSize * 2}px`,
  xl: `${STYLE_BASE.fontSize * 3}px`,
};

export const MAX_SPACE_UNIT_MULTIPLE = 20;

/**
 * returns an object of spacing units for things like padding/margin
 * @param {Number} maxSpacingMultiple the maximum multiple of the spacing factor
 * @param {Number} spacingFactor
 * @returns {Object} => {'1x': '8px', '2x': '16px'};
 */
const getSpacingUnits = (maxSpacingMultiple, spacingFactor) => {
  const spacingUnits = {};
  for (let i = 1; i <= maxSpacingMultiple; i++) {
    spacingUnits[`${i}x`] = `${i * spacingFactor}px`;
  }

  return spacingUnits;
};

// creates a object that has multiples of spacing units in pixels
// SPACING['1x'] => '5px'
export const SPACING = getSpacingUnits(MAX_SPACE_UNIT_MULTIPLE, STYLE_BASE.spacing);

// bootstrap breakpoints used for emotion custom styled components
export const BOOTSTRAP_BREAKPOINTS = [
  { alias: 'sm', width: 576 },
  { alias: 'md', width: 768 },
  { alias: 'lg', width: 992 },
  { alias: 'xl', width: 1200 },
];

const breakpoint = (amount, unit = 'px') => `@media (min-width: ${amount}${unit})`;
/**
 * returns an object aliased by bootstrap breakpoint values
 * {
 *   sm: @media (min-width: ...)
 *   md: ...
 * }
 * Usage:
 * https://emotion.sh/docs/media-queries#reusable-media-queries
 */
export const EMOTION_BOOTSTRAP_BREAKPOINTS = BOOTSTRAP_BREAKPOINTS.reduce(
  (breakpoints, currentBp) => {
    breakpoints[currentBp.alias] = breakpoint(currentBp.width);
    return breakpoints;
  },
  {},
);

export const CUSTOM_BREAKPOINTS = {
  navbar: breakpoint(932),
};

// COLOR PALETTE TOKENS
// We have two palettes that are in use
// the design-system palette https://developer.gov.bc.ca/Design-System/Colour-Palette
// and our custom devhub palette for further customization

// sample usage would be  (using emotion css)
// import { DS_PALETTE } from '...'
// css` border-color: ${DS_PALETTE.brand.yellow}`;
// bc design-system palette
export const DS_PALETTE = {
  brand: {
    yellow: '#FCBA19',
    blue: '#003366',
  },
  text: {
    darkgrey: '#494949',
  },
  links: {
    blue: '#1A5A96',
  },
  backgrounds: {
    blue: '#38598A',
    lightgrey: '#F2F2F2',
  },
  red: '#D8292F',
  green: '#2E8540',
};

// devhub specific palette
export const DEVHUB_PALETTE = {
  primary: DS_PALETTE.brand.blue,
  secondary: DS_PALETTE.brand.yellow,
  [RESOURCE_TYPES.COMPONENTS]: '#4299D2',
  [RESOURCE_TYPES.DOCUMENTATION]: '#246BD1',
  [RESOURCE_TYPES.PEOPLE]: '#E17039',
  [RESOURCE_TYPES.SELF_SERVICE_TOOLS]: '#F3BA45',
  [TOPICS]: '#444',
  [RESOURCE_TYPES.REPOSITORIES]: '#6e5494',
  [SEARCH_RESOURCE_TYPES.GITHUB_ISSUE]: '#6e5494',
  [RESOURCE_TYPES.EVENTS]: '#f44b42',
  link: DS_PALETTE.links.blue,
};
