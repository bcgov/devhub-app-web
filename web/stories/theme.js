import { create } from '@storybook/theming';
import { DEVHUB_PALETTE } from '../theme';

export default create({
  base: 'light',

  colorPrimary: 'hotpink',
  colorSecondary: 'deepskyblue',

  // UI
  appBg: 'white',
  appContentBg: 'white',
  appBorderColor: 'grey',
  appBorderRadius: 4,

  // Typography
  fontBase: 'sans-serif',
  fontCode: 'monospace',

  // Text colors
  textColor: DEVHUB_PALETTE.darkgrey,
  textInverseColor: 'rgba(255,255,255,0.9)',

  // Toolbar default and active colors
  barTextColor: DEVHUB_PALETTE.white,
  barSelectedColor: 'black',
  barBg: DEVHUB_PALETTE.blue,

  // Form colors
  inputBg: 'white',
  inputBorder: 'silver',
  inputTextColor: 'black',
  inputBorderRadius: 4,

  brandTitle: 'My custom storybook',
  brandImage:
    'https://u58tqm22fk-flywheel.netdna-ssl.com/wp-content/uploads/2016/06/BC-Gov-Logo-CMYK_pos.png',
});
