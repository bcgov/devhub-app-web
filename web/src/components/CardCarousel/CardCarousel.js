import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import Carousel from '@brainhubeu/react-carousel';
import '@brainhubeu/react-carousel/lib/style.css';
import { faChevronRight, faChevronLeft } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { EMOTION_BOOTSTRAP_BREAKPOINTS } from '../../constants/designTokens';
import { RESOURCE_TYPES_LIST, CARD_CAROUSEL } from '../../constants/ui';
import Card from '../Card/Card';

export const TEST_IDS = {
  arrowLeft: 'carousel-arrow-left',
  arrowRight: 'carousel-arrow-right',
};

const Icon = styled(FontAwesomeIcon)`
  display: none;
  ${EMOTION_BOOTSTRAP_BREAKPOINTS.sm} {
    color: ${({ theme }) => theme.colors.primary};
    font-size: 50px;
    padding: 0 5px;
    cursor: pointer;
    display: block;
  }
`;

/**
 * returns whether or not the arrow buttons should show in the carousel
 * based on certain cases the arrow components are returned or not returned
 * @param {Number} numResources the amount of resources within the carousel
 * @param {Number} index the current page position
 * @param {Number} slidesPerPage the amount of slides being displayed in the carousel
 * @returns {Object} the arrow props (or not) to be merged into the remaining carousel props
 */
const showArrow = (numResources, index, slidesPerPage) => {
  if (numResources <= slidesPerPage) {
    return {};
  }
  const maxPages = Math.ceil(numResources / slidesPerPage);
  const nominalSlides = maxPages * slidesPerPage;
  return {
    arrowLeft: index === 0 ? null : <Icon data-testid={TEST_IDS.arrowLeft} icon={faChevronLeft} />,
    arrowRight:
      index + slidesPerPage >= nominalSlides ? null : (
        <Icon data-testid={TEST_IDS.arrowRight} icon={faChevronRight} />
      ),
  };
};

const CardCarousel = ({ resources }) => {
  const [index, setIndex] = useState(0);
  const numResources = resources.length;

  const slides = resources.map(r => (
    <Card
      resourceType={r.fields.resourceType}
      key={r.id}
      title={r.fields.title}
      description={r.fields.description}
      image={r.fields.image}
      link={r.path}
      event={r}
    />
  ));

  const settings = {
    animationSpeed: 125,
    value: index,
    onChange: setIndex,
    ...showArrow(numResources, index, CARD_CAROUSEL.desktop.slidesPerPage),
    addArrowClickHandler: true,
    slidesPerPage: CARD_CAROUSEL.desktop.slidesPerPage,
    slidesPerScroll: CARD_CAROUSEL.desktop.slidesPerScroll,
    itemWidth: 275,
    breakpoints: {
      500: {
        slidesPerPage: CARD_CAROUSEL.mobile.slidesPerPage,
        slidesPerScroll: CARD_CAROUSEL.mobile.slidesPerScroll,
      },
    },
  };
  return <Carousel {...settings}>{slides}</Carousel>;
};

CardCarousel.propTypes = {
  resources: PropTypes.arrayOf(
    PropTypes.shape({
      type: PropTypes.oneOf(RESOURCE_TYPES_LIST),
      unfurl: PropTypes.shape({
        title: PropTypes.string.isRequired,
        description: PropTypes.string,
        image: PropTypes.string,
      }),
      resource: PropTypes.shape({
        path: PropTypes.string.isRequired,
        type: PropTypes.oneOf(RESOURCE_TYPES_LIST).isRequired,
      }),
    }),
  ),
};

export default CardCarousel;
