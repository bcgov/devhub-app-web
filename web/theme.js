/*
Copyright 2019 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/

/*
Emotion theming object

We are transitioning from common elements such as colors, background colors being abstracted away into
a themes provider. This is a very common practice now a days and makes it maintain themes on the fly
docs on theming is here https://emotion.sh/docs/theming#___gatsby

Docs on emotion https://emotion.sh/docs/introduction

--- our idea is to introduce emotion early and slowly replace css modules with emotion styled components
when and as needed. This is a slow processes but will lead to better and more maintaible styling in the long run.
*/

import { SPACING, DEVHUB_PALETTE } from './src/constants/designTokens';
export default {
  colors: {
    ...DEVHUB_PALETTE,
  },
  padding: {
    main: {
      mobile: `${SPACING['2x']} ${SPACING['2x']}`,
      desktop: `${SPACING['2x']} ${SPACING['13x']}`,
    },
  },
  breakpoints: {
    // specific breakpoints
    main: {
      desktop: '@media screen and (min-width: 932px)',
    },
  },
};
