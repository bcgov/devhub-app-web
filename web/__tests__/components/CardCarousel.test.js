import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import { ThemeProvider } from '@emotion/react';
import theme from '../../theme';
import CardCarousel, { TEST_IDS } from '../../src/components/CardCarousel/CardCarousel';
import { RESOURCE_TYPES } from '../../src/constants/ui';
// mock out card components
jest.mock('../../src/components/Card/Card.js', () => () => <div>card</div>);

describe('Card Carousel', () => {
  afterEach(cleanup);

  const resources = [
    {
      id: 'foo',
      resourceType: RESOURCE_TYPES.DOCUMENTATION,
      fields: {
        resourceType: RESOURCE_TYPES.DOCUMENTATION,
        title: 'foo',
        description: 'bar',
        standAlonePath: '/foo',
      },
      path: '/foo',
    },
  ];

  it('matches snapshot', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <CardCarousel resources={resources} />
      </ThemeProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('does not show any arrows if there are not enough components to go make another page', () => {
    const { queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <CardCarousel resources={resources} />
      </ThemeProvider>,
    );

    expect(queryByTestId(TEST_IDS.arrowLeft)).not.toBeInTheDocument();
    expect(queryByTestId(TEST_IDS.arrowRight)).not.toBeInTheDocument();
  });

  it('shows the right arrow if there are more than 3 resources when clicked the right arrow disappears since there are no more pages to goto and the left one shows up', () => {
    const { queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <CardCarousel
          resources={resources
            .concat(resources)
            .concat(resources)
            .concat(resources)}
        />
      </ThemeProvider>,
    );

    const RightArrow = queryByTestId(TEST_IDS.arrowRight);
    const LeftArrow = queryByTestId(TEST_IDS.arrowLeft);

    expect(LeftArrow).not.toBeInTheDocument();
    expect(RightArrow).toBeInTheDocument();

    fireEvent.click(RightArrow);
    // wait for re renders to happen
    expect(queryByTestId(TEST_IDS.arrowRight)).not.toBeInTheDocument();
    expect(queryByTestId(TEST_IDS.arrowLeft)).toBeInTheDocument();
  });
});
