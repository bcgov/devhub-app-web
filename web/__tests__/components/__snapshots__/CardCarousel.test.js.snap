// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Carousel matches snapshot 1`] = `
<div>
  <div
    class="BrainhubCarousel"
  >
    <div
      class="BrainhubCarousel__trackContainer"
    >
      <ul
        class="BrainhubCarousel__track"
        style="width: 0px; margin-left: 0px; transform: translateX(0px);"
      >
        <li
          class="BrainhubCarouselItem BrainhubCarouselItem--active"
          style="padding-right: 0px; padding-left: 0px; width: 275px; max-width: 275px; min-width: 275px;"
        >
          <div>
            card
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
